astropy/wcs/utils.py

```diff
@@ -1137,3 +1141,47 @@ def fit_wcs_from_points(xy, world_coords, proj_point='center',
                       np.zeros((degree+1, degree+1)), wcs.wcs.crpix)
 
     return wcs
+
+
+def obsgeo_to_frame(obsgeo, obstime):
+    """
+    Convert a WCS obsgeo property into an ITRS coordinate frame.
+
+    Parameters
+    ----------
+    obsgeo : np.ndarray
+        A shape ``(6, )`` array representing ``OBSGEO-[XYZ], OBSGEO-[BLH]`` as
+        returned by `astropy.wcs.WCS.wcs.obsgeo`.
+
+    obstime : time-like
+        The time assiociated with the coordinate, will be passed to
+        `astropy.coordinates.ITRS` as the obstime keyword.
+
+    Returns
+    -------
+    BaseCoordinateFrame
+        An `~astropy.coordinates.ITRS` coordinate frame representing the coordinates.
+    """
+
+    if (obsgeo is None
+        or np.all(obsgeo == 0)
+        or np.all(~np.isfinite(obsgeo))
+    ):
+        raise ValueError(f"Can not parse the 'obsgeo' location ({obsgeo}). "
+                         "obsgeo should be non-zero, finite and not None."
+                         )
+
+    # If the cartesian coords are not zero (should only be false if .set() hasn't been called.
+    if np.all(obsgeo[:3] != 0):
+        data = CartesianRepresentation(*obsgeo[:3] * u.m)
+
+    # If the cartesian coords are all zero then use the spherical ones
+    elif np.all(obsgeo[3:] != 0):
+        data = SphericalRepresentation(*[comp * unit for comp, unit in zip(obsgeo[3:], (u.deg, u.deg, u.m))])
+
+    else:
+        raise ValueError(f"Could not parse a valid set of coordinates from the obsgeo struct {obsgeo} "
+                         "All of either the XYZ or LBH values should be non-zero. "
+                         "Have you tried running `wcs.set()`?")
```

Comment:
This is something I am not familiar about [... conventions in `astropy`]: Do you use backticks in exception messages? Or should they be used? [It's just a question. no need to modify if it seems OK]

Reply:
Yeah, probably best not to, they don't get formatted anywhere, so I removed them.